src/svd/util.py

```diff
@@ -0,0 +1,386 @@
+                        periph_id = _get_periph_id(periph)
+                        if periph_id in reg_map:
+                            current_periph_regs = reg_map[periph_id]
+                        else:
+                            current_periph_regs = self._periph_reg_map(current_periph)
+                            reg_map[periph_id] = current_periph_regs
+                        self._written_peripherals.setdefault(periph_id, periph)
+                        break
+                else:
+                    svd.log.warning(
+                        f"Address 0x{addr_0:08x} does not correspond to any peripheral"
+                    )
+                    continue
+
+            assert current_periph_regs is not None
+
+            try:
+                reg = current_periph_regs[addr_0]
+            except KeyError:
+                svd.log.warning(
+                    f"Address 0x{addr_0:08x} is within the address range of {current_periph} "
+                    f"[0x{current_periph_range.start:x}-0x{current_periph_range.stop:x}), but "
+                    "does not correspond to any register in the peripheral"
+                )
+                continue
+
+            reg_len = reg.bit_width // 8
+            reg_content_bytes = [val_0]
+            for i in range(reg_len - 1):
+                try:
+                    _, val_i = next(map_iter)
+                    reg_content_bytes.append(val_i)
+                except StopIteration:
+                    raise ValueError(
+                        f"Content for {reg} was only partially specified. "
+                        f"Missing value for address 0x{addr_0 + i:08x}"
+                    )
+
+            # TODO: don't need to call this more than once
+            if not self._enforce_svd_constraints:
+                reg.unconstrain()
+
+            reg.content = int.from_bytes(bytes(reg_content_bytes), byteorder="little")
+
+        return self
+
+    def apply_dict(self, config: dict[str, Any]) -> DeviceBuilder:
+        """Populate device content from a dictionary representation of the registers and content.
+
+        The dictionary structure should match the structure of the device peripherals and registers.
+        Content can be set either at the register or field level.
+        Field content can be set either using an enum name or with a numeric value.
+
+        :param content: content dictionary.
+        :returns: builder instance.
+        """
+        affected_periphs = []
+
+        for periph_name, content in config.items():
+            peripheral = self._device[periph_name]
+            for reg_name, reg_value in content.items():
+                self._reg_apply_dict(
+                    peripheral[reg_name],
+                    reg_value,
+                )
+            affected_periphs.append(peripheral)
+
+        for periph in affected_periphs:
+            self._written_peripherals.setdefault(_get_periph_id(periph), periph)
```

Comment:
No need for the intermediate list.
```suggestion
            self._written_peripherals.setdefault(_get_periph_id(peripheral), peripheral)
```